From a057e87878ab8ad1e742ca2d3f0704d89cb098e7 Mon Sep 17 00:00:00 2001
From: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
Date: Mon, 23 Jul 2018 13:56:45 -0500
Subject: [PATCH 1814/2302] ASoC: SOF: trace: fix linker error in i386 mode

64-bit division or modulo needs to rely on do_div, cannot use / or %
in 32-bit mode

Fixes linker error

ERROR: "__moddi3" [sound/soc/sof/snd-sof.ko] undefined!
scripts/Makefile.modpost:92: recipe for target '__modpost' failed

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
---
 sound/soc/sof/trace.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/sound/soc/sof/trace.c b/sound/soc/sof/trace.c
index 04acd78..1b37970 100644
--- a/sound/soc/sof/trace.c
+++ b/sound/soc/sof/trace.c
@@ -73,6 +73,7 @@ static ssize_t sof_dfsentry_trace_read(struct file *file, char __user *buffer,
 	unsigned long rem;
 	loff_t lpos = *ppos;
 	size_t avail, buffer_size = dfse->size;
+	u64 lpos_64;
 
 	/* make sure we know about any failures on the DSP side */
 	sdev->dtrace_error = false;
@@ -84,7 +85,9 @@ static ssize_t sof_dfsentry_trace_read(struct file *file, char __user *buffer,
 		return 0;
 
 	/* check for buffer wrap and count overflow */
-	lpos = lpos % buffer_size;
+	lpos_64 = lpos;
+	lpos = do_div(lpos_64, buffer_size);
+
 	if (count > buffer_size - lpos)
 		count = buffer_size - lpos;
 
-- 
2.7.4

